### PR TITLE
Check dataset for metrics

### PR DESCRIPTION
**Pull request by b0nce**

### CODE REVIEW

File: tests/test_oml/test_metrics/test_embedding_metrics.py [hunk 1]
```diff
@@ -206,3 +207,63 @@ def test_mixed_epochs(perfect_case, imperfect_case, worst_case):  # type: ignore
...
+    # 0, 1, 1, 0, 1
+    isg = torch.zeros(5)
+    isg[[1, 2, 4]] = 1
+
+    calc.acc.storage[calc.labels_key] = l
+    calc.acc.storage[calc.is_query_key] = isq
+    calc.acc.storage[calc.is_gallery_key] = isg
+
+    # is_storage_full will return False in this case
+    calc.acc.num_samples  # type: ignore
```
Comment (AlekseySh) [inline tests/test_oml/test_metrics/test_embedding_metrics.py:240]: looks like nothing happens here

File: tests/test_oml/test_metrics/test_embedding_metrics.py [hunk 2]
```diff
@@ -206,3 +207,63 @@ def test_mixed_epochs(perfect_case, imperfect_case, worst_case):  # type: ignore
...
+
+    # 1, 0, 0, 1, 1
+    isq = torch.zeros(5)
+    isq[[0, 3, 4]] = 1
+
+    # 0, 1, 1, 0, 1
+    isg = torch.zeros(5)
+    isg[[1, 2, 4]] = 1
+
+    calc.acc.storage[calc.labels_key] = l
```
Comment (AlekseySh) [inline tests/test_oml/test_metrics/test_embedding_metrics.py:235]: can you use `.update_data` instead? it's a normal way to fill the storage

File: tests/test_oml/test_metrics/test_embedding_metrics.py [hunk 3]
```diff
@@ -206,3 +207,63 @@ def test_mixed_epochs(perfect_case, imperfect_case, worst_case):  # type: ignore
...
+    calc.acc.storage[calc.is_query_key] = isq
+    calc.acc.storage[calc.is_gallery_key] = isg
+
+    # is_storage_full will return False in this case
+    calc.acc.num_samples  # type: ignore
+    calc.acc._collected_samples = 3
```
Comment (AlekseySh) [inline tests/test_oml/test_metrics/test_embedding_metrics.py:241]: with `.update_data` you don't need this

File: tests/test_oml/test_metrics/test_embedding_metrics.py [hunk 4]
```diff
@@ -206,3 +207,63 @@ def test_mixed_epochs(perfect_case, imperfect_case, worst_case):  # type: ignore
...
+    with pytest.raises(ValueError):
+        # fails because of is_storage_full()
+        calc._validate_dataset()
+    calc.acc.num_samples = 3  # type: ignore
```
Comment (AlekseySh) [inline tests/test_oml/test_metrics/test_embedding_metrics.py:245]: with `.update_data` you don't need this

File: tests/test_oml/test_metrics/test_embedding_metrics.py [hunk 5]
```diff
@@ -206,3 +207,63 @@ def test_mixed_epochs(perfect_case, imperfect_case, worst_case):  # type: ignore
...
+
+    with pytest.raises(ValueError):
+        # fails because there is no mask_to_ignore
+        calc._validate_dataset()
+    ids_query = torch.nonzero(isq).squeeze()
+    ids_gallery = torch.nonzero(isg).squeeze()
+    calc.mask_to_ignore = ids_query[..., None] == ids_gallery[None, ...]
```
Comment (AlekseySh) [inline tests/test_oml/test_metrics/test_embedding_metrics.py:252]: we should not repeat the internal logic of the class in the test

File: tests/test_oml/test_metrics/test_embedding_metrics.py [hunk 6]
```diff
@@ -206,3 +207,63 @@ def test_mixed_epochs(perfect_case, imperfect_case, worst_case):  # type: ignore
...
+
+    with pytest.raises(ValueError):
+        # fails because there is no mask_gt
+        calc._validate_dataset()
+    calc.mask_gt = calc_gt_mask(labels=l, is_query=isq, is_gallery=isg)
```
Comment (AlekseySh) [inline tests/test_oml/test_metrics/test_embedding_metrics.py:257]: the same

File: tests/test_oml/test_metrics/test_embedding_metrics.py [hunk 7]
```diff
@@ -206,3 +207,63 @@ def test_mixed_epochs(perfect_case, imperfect_case, worst_case):  # type: ignore
...
+    with pytest.raises(AssertionError):
+        # fails because of bad queries
+        calc._validate_dataset()
+
+    # 1, 0, 0, 1, 0
+    isq[-1] = 0
+    ids_query = torch.nonzero(isq).squeeze()
+    ids_gallery = torch.nonzero(isg).squeeze()
+    calc.mask_to_ignore = ids_query[..., None] == ids_gallery[None, ...]
+    calc.mask_gt = calc_gt_mask(labels=l, is_query=isq, is_gallery=isg)
```
Comment (AlekseySh) [inline tests/test_oml/test_metrics/test_embedding_metrics.py:268]: same

**Review by AlekseySh — CHANGES_REQUESTED** [review]
The function itself looks good, but tests must not repeat the internal logic of `EmbeddingMetrics` since it may easily change in the future

It's better to use the provided interface of ` EmbeddingMetrics`:
```
calc =  EmbeddingMetrics()
clc.setup()
calc.update_data()
calc.calculate_metrics()
```

File: tests/test_oml/test_metrics/test_embedding_metrics.py (hunk 1 above)
Comment (b0nce) [reply tests/test_oml/test_metrics/test_embedding_metrics.py:240]: wdym?

File: tests/test_oml/test_metrics/test_embedding_metrics.py (hunk 3 above)
Comment (b0nce) [reply tests/test_oml/test_metrics/test_embedding_metrics.py:241]: same here

File: tests/test_oml/test_metrics/test_embedding_metrics.py (hunk 4 above)
Comment (b0nce) [reply tests/test_oml/test_metrics/test_embedding_metrics.py:245]: again same here

File: tests/test_oml/test_metrics/test_embedding_metrics.py (hunk 5 above)
Comment (b0nce) [reply tests/test_oml/test_metrics/test_embedding_metrics.py:252]: yes, but as you said earlier you don't know how we will implement mask_to_ignore calculation and we will still have to change this test

Comment (b0nce) [reply tests/test_oml/test_metrics/test_embedding_metrics.py:252]: at most I can suggest to use `.calculate_mask_to_ignore()` as a class method for now

File: oml/functional/metrics.py [hunk 8]
```diff
@@ -189,6 +193,10 @@ def calculate_accuracy_on_triplets(embeddings: torch.Tensor, reduce_mean: bool =
         return acc
 
 
+def validate_dataset(mask_gt: torch.Tensor, mask_to_ignore: torch.Tensor) -> None:
+    assert (mask_gt & ~mask_to_ignore).any(1).all(), "There are queries without galleries!"  # type: ignore
```
Comment (AlekseySh) [inline oml/functional/metrics.py:197]: "There are queries without available correct answers in the gallery!

**Review by AlekseySh — APPROVED** [review]

**Pull request merged**